lib/asnyc-promise.js

```diff
@@ -0,0 +1,20 @@
+var resolveAsync = require('./async');
+
+function resolve(x, options, callback) {
+    // backward compatibility for using callbacks
+    if (typeof options === 'function' || typeof callback === 'function') {
+        return resolveAsync(x, options, callback);
+    }
+    // return a promise in other cases
+    /* global Promise */
+    return new Promise(function (promiseResolve) {
```

Comment:
`util.promisify` exists; what's the value of adding this API here?

Reply:
It shows that Broswerify actually cares about modernizing its packages. Otherwise, it gives the wrong message to the users. It also gives you a chance to start using `async/awit`.

Reply:
lol what? "modernizing" doesn't require using promises; callbacks are still a thing.

You can `await util.promisify(require('resolve'))(...args)` just fine right now.

Reply:
> lol what? "modernizing" doesn't require using promises; callbacks are still a thing.

Unless you want to stick with the pyramid of doom that is caused by callback hell. 😄😉 

> You can await util.promisify(require('resolve'))(...args) just fine right now.

So you are suggesting that I make another one-liner npm package instead of contributing to the original package?

Reply:
No, I'm suggesting you do `const resolve = require('util').promisify(require('resolve'));` in whatever file you need rather than needing anything upstream.

Adding this here increases the complexity for all `resolve` uysers; it introduces an implicit dependency on Promise, which isn't present in node < 0.12 (this package supports down to node 0.4), and causes the currently explicit "only an async callback" behavior to become a hybrid, implicit "maybe it returns a Promise, maybe not", which is commonly referred to as [z̲̗̼͙̥͚͛͑̏a̦̟̳͋̄̅ͬ̌͒͟ļ̟̉͌ͪ͌̃̚g͔͇̯̜ͬ̒́o̢̹ͧͥͪͬ](https://blog.izs.me/2013/08/designing-apis-for-asynchrony). At the very least, this would need to always, unconditionally, return a Promise, whether the user provided a callback or not.

Additionally, this code is incorrect as written, because `resolve` calls the async callback with *more than one success argument*, so since Promises can only hold one value, the Promise form would have to be some sort of array or object container - which means the callback and Promise APIs would have inconsistent signatures, which simply isn't an acceptable state to me.

Reply:
I'm already using my function somewhere, but since I use `resolve` in multiple places, it suggests that I should make yet another one-liner npm package.

> Adding this here increases the complexity for all resolve uysers; it introduces an implicit dependency on Promise, which isn't present in node < 0.12 (this package supports down to node 0.4)

That's exactly what I meant by the wrong message in the above comments. It seems that the goal of Browserify is to fulfill technical debts and support legacy code and not modernizing code.

> , and causes the currently explicit "only an async callback" behavior to become a hybrid, implicit "maybe it returns a Promise, maybe not", which is commonly referred to as z̲̗̼͙̥͚͛͑̏a̦̟̳͋̄̅ͬ̌͒͟ļ̟̉͌ͪ͌̃̚g͔͇̯̜ͬ̒́o̢̹ͧͥͪͬ. At the very least, this would need to always, unconditionally, return a Promise, whether the user provided a callback or not.
> Additionally, this code is incorrect as written, because resolve calls the async callback with more than one success argument, so since Promises can only hold one value, the Promise form would have to be some sort of array or object container - which means the callback and Promise APIs would have inconsistent signatures, which simply isn't an acceptable state to me.

The purpose is to not use callbacks, not to return a promise when someone uses a callback. 

Structured concurrency is a well-known phenomenon that is used in other languages too. I don't quite understand the point of the vaguely written article you linked here. If you want a scientific discussion about the subject, you can take a look at this article:

https://ericniebler.com/2020/11/08/structured-concurrency

Reply:
There seem to be some misunderstandings here.

"Modernizing" is not necessarily - or even often - a good thing. Things aren't better just because they're newer. Callbacks aren't inherently bad, and "callback hell" is a largely fictitious problem that has always been trivially solved by function abstractions and intermediate variables. Promises are wonderful, and I love them, but "callback bad Promise good" is simply a false ideology.

More importantly though, is [maintaining backwards compatibility](https://steve-yegge.medium.com/dear-google-cloud-your-deprecation-policy-is-killing-you-ee7525dc05dc). The primary message any user of EVERY PACKAGE I MAINTAIN receives should always be, _it will work for as close to forever as is feasible_. This doesn't mean _no_ breaking changes, but it does mean that breaking changes are minimal, highly motivated, and not capricious. It also means, since I'm committing to maintain everything for eternity, that any single change or new feature has a potentially very high cost, and has to be weighed accordingly.

If you think that saving you the need to write a line or three of code, or, to write a three-line package wrapper, is worth imposing extra complexity on a package that's downloaded 35 million times every week, then you need to make a very strong case for it, and "people want SHINY" is not a compelling argument, nor is implying that a package is somehow "backwards" for not enthusiastically accepting your ~~feature~~ higher maintenance cost.